### DISCUSSION TITLE
Allow additional hook names with Eslint Plugin Query

### DISCUSSION THREAD

**Discussion by thespacemanatee** [TanStack/query#8025] [Ideas] [answered]
Thank you for this wonderful library.

We are using this library in React Native and we are wrapping `useQuery` with some React Navigation stuff to get `refetchOnWindowFocus` working. However, the Eslint Plugin Query does not recognise these custom wrappers as Query hooks and does not validate them.

We would like to see an `additionalHooks` setting on the rules, similar to how React does it:

```js
{
  "rules": {
    // ...
    "react-hooks/exhaustive-deps": ["warn", {
      "additionalHooks": "(useMyCustomHook|useMyOtherCustomHook)"
    }]
  }
}
```

Thank you very much for considering this feature request!

**Comment by TkDodo** [comment]
@Newbie012 what do you think ?

**Reply by Newbie012** [reply]
IIRC the way the rule works is by looking at objects containing both `queryKey` and `queryFn`, and the function that wraps them is imported from tanstack. This way, the rule doesn't have to worry if someone imported `{ useQuery as useReactQuery }` or imported `useQuery` from other library, and as a result, newly introduced functions (such as `createQuery` and `queryOptions`) were supported out of the box.

I guess we could add the option to support custom hooks (without imports check), while keeping current behavior. This way, the responsibility of conflicts will be passed to the consumer.

thoughts?

**Reply by TkDodo** [reply]
Sounds good 👍

**Reply by Newbie012** [reply, marked as answer]
Apparently I don't check the imports. It's already working as long as an object contains both `queryKey` and `queryFn`: https://eslint-online-playground.netlify.app/#eNqNkEFrwzAMhf+K8CUtpAnbMaNj0LHLTjvXPYREKe5iOY3t0RLy3yfbZWQwRk+W0XufnjQJOzYlXmo99FicrKhE56lxyhDsjB4MIbnVBKrNwVv88DheY7Xz1hkd/zCvYZIE0BiyDs4PsP3RsvUc3ne8VrA/5On3RhWs1rB9Zi67nxbmx2Re4O9FzCIXaHtFrmBWp45pHcVLjA6G3h8VpbzdaDRkL67miXXzWSbXJkk2EZ9JkoSXaG2xq33vYB9iFkWxQN0m2X3W9bUrR2yM1kgtttkhl3SQxKEGnlEfw3UNcaB4KincdUApKq60aX3PNRtCo8WvVxwChBqFNmiiJfT+zZxonAOtu9GiJ0nLU0T9IUj9383Qm8NN528O0brm

**Reply by thespacemanatee** [reply]
> Apparently I don't check the imports. It's already working as long as an object contains both `queryKey` and `queryFn`: https://eslint-online-playground.netlify.app/#eNqNkEFrwzAMhf+K8CUtpAnbMaNj0LHLTjvXPYREKe5iOY3t0RLy3yfbZWQwRk+W0XufnjQJOzYlXmo99FicrKhE56lxyhDsjB4MIbnVBKrNwVv88DheY7Xz1hkd/zCvYZIE0BiyDs4PsP3RsvUc3ne8VrA/5On3RhWs1rB9Zi67nxbmx2Re4O9FzCIXaHtFrmBWp45pHcVLjA6G3h8VpbzdaDRkL67miXXzWSbXJkk2EZ9JkoSXaG2xq33vYB9iFkWxQN0m2X3W9bUrR2yM1kgtttkhl3SQxKEGnlEfw3UNcaB4KincdUApKq60aX3PNRtCo8WvVxwChBqFNmiiJfT+zZxonAOtu9GiJ0nLU0T9IUj9383Qm8NN528O0brm

I just tried in my setup and it doesn't work if the wrapper has a different name like `useFocusEffectQuery` and isn't exported from the official library.

EDIT: Saw the playground and I agree that there's no reason it shouldn't work, I'll take a look at my setup again, thanks!

**Reply by kennygrubb** [reply]
Some rules, like no-rest-destructuring, do not fire in this scenario. 

A common use-case is using the TRPC react-query wrappers ... would really like to be able to use this eslint plugin with it.

**Reply by Newbie012** [reply]
I'm not sure what you mean. This rule applies in other syntaxes such as `trpc.hello.useQuery({ queryKey: [], queryFn: () => id })`.

**Reply by kennygrubb** [reply]
TRPC handles internally the queryKey and queryFn, so those are not required when calling useQuery

https://trpc.io/docs/client/react/useQuery

**Reply by dd-jonas** [reply]
@Newbie012 Can we get this to work when using custom query hooks?

```ts
const useInvoicesQuery = () => useQuery({ ... });

const Component = () => {
    const query = useInvoicesQuery();
};
```

I feel like this is a very common pattern, and the plugin does not recognize these hooks.

**Reply by Newbie012** [reply]
@dd-jonas what do you mean? this is working as expected

https://eslint-online-playground.netlify.app/#eNp9kUFrhDAQhf/KkItb2NW7y5ZCS6H01LPxIDpKtjqxSVxWxP/eSWKhLcuejLz3vnmTLMKaOsNrNYw9pmcrclFrsg4mi2900apG+zGhmeEEuwVUA+sDnB5hkQRg0E2GvDVYWP/y33eccyjKffx7pRx2IRPCR0mrJElxyrMeRk1I7hYeIJoChg3/K/0EPPIo9gJtr8ilHGpVF3dRjDcOxn7qFMU9WqMHSJ5cxeiq/sxi6hAthzAr8QXxGqINttXUOyh8ozRNf6G2SbZI2r5ymcFaDwNSg01S7iWVkrjUyDOqzl+tJi4UFpPCzSNKkfNp0M3U85kDXmjw8oKjh1Ct0HrPdhdS3O0cadwDrdtoIROt2Tmgbhii/lf0Gj/SKtZvO7m51A==

**Reply by dd-jonas** [reply]
Thanks for the quick reply @Newbie012. I'm specifically talking about the rules `@tanstack/query/no-rest-destructuring` and `@tanstack/query/no-unstable-deps`.

For some reason, I can't trigger a rest destructure error in the playground, even when copying the incorrect example from the docs. Here's a different example with a mutation:
https://eslint-online-playground.netlify.app/#eNqVkkFP4zAQhf/KyJcEqSTaHrtitRK7K+0BJLg2PRhnUgLOODh2Bary3xnbaWmhIHEbZd6872U8WzFYVeKz7HqNxcMgFqLtemMdbMEPeCm1vpPqEUZorOkgu0WpXFbRoejKO+laQ3vRbydpcDxW2iA/f/JoX3ioImW4EY254/A/bUyrEC4gP4OLX4dm+Ra6qfxHC8g3UnuMoljBePYzGDaeVGRfGk5ESI6dthUBJNTO4wczvuv+3mOePI6i5ynGTqqmfU243fry9HsxF7xlKmKFeXaPWpssMcsS/lprbKjHGSz36tVp1C7Vl6j5Z6hrA3iKNt/TLDpvCchrzV/GisRM4KBbcgUHadr10dX02q9bugkP/uEY0tR5khwcBT7H0Rob6bWDZaAWRXFgNZGGZdZo6fiqlOk6pBrrbDWraBVD9cyQ63DEhjhQ3EAl3EuPlVhw1Znaa655IDRq3PzBPpiQanEImmlplfgyc3LjHDi4yS3OJGn5EK1OCFL/uBnXzlsV4ytdWTdl

**Reply by Newbie012** [reply]
@TkDodo may have a better idea for this specific use case. Since `mutation1` is indeed unstable dependency, its `mutate` method is stable. So what about this one?

https://eslint-online-playground.netlify.app/#eNqVkkFP4zAQhf/KyJcEqSTaHrtitRK7SBxAgmvTg3EmJeCMg2NXoCr/nbGdQlsqJG6j+L3vvYxmKwarSnyVXa+xeBrEQrRdb6yDLfgBL6XWD1I9wwiNNR1k9yiVyyraF914J11r6EP010kaHNtKG+TnLx7tG5sqUoYfIphfHF7TxrQK4QLyM7j4sw/Lt9BN4xUtIN9I7TGK4gTj2e8AbDypmH1puBEhOSZtKwJIUTvGL874Kf2YMU+Mg+p5qrGTqmlfU9xufXn6vdgLPjsVccIiuHLyWs8ge0StTZbiyxL+W2tsmMcZLI+Nq9Phu57fhs8nRn6ceGsAT4XOP9IsOm8JQmH+MlYkZgIH3ZIruEjTrg/uqNd+3dJdOIEv55Fc50mydyb4Gq01NtJrB8uQWhTFHmpKGpZZo6XjO1Om65BqrLPVrKJVLNVzhlyHszbEheIGKuHeeqzEgqfO1F7zzIbwUOPmH/YBQqrFIWimpVXi286Jxj1wcBMtepK0fIqoE4L0fvgY185bFeM7WMo99A==

**Reply by TkDodo** [reply]
if you need to put `mutate` into a dependency array, yes, it has to be with `mutation1.mutate` directly.

the issue by @dd-jonas seems like a limitation of `no-unstable-deps`, as it likely only checks for values coming from `useMutation` directly, not from custom hooks like `useCreateInvoice`. Instead of adding additional hook names (which can be quite many), maybe there’s a way to make the rule catch those, too? Not sure ...

**Reply by Newbie012** [reply]
Oh now I get it. Yes, this is possible by adjusting the rule to check the hook by its return type rather than its name. The only downside with this solution is that it will require type checking which **may** have a minor performance cost.

**Reply by TkDodo** [reply]
Do we already require type checking for any of our rules? Is it possible to have the rule work without it for users who don’t use TypeScript, or would this need a separate rule?
